File: from-2601-to-2700/2658-maximum-number-of-fish-in-a-grid/solution.go
func findMaxFish(grid [][]int) int {
    m, n := len(grid), len(grid[0])
    var dfs func(r, c int) int
    dfs = func(r, c int) int {
        if r < 0 || r >= m || c < 0 || c >= n || grid[r][c] == 0 {
            return 0
        }
        fish := grid[r][c]
        grid[r][c] = 0
        fish += dfs(r + 1, c)
        fish += dfs(r - 1, c)
        fish += dfs(r, c + 1)
        fish += dfs(r, c - 1)
        return fish
    }
    maxFish := 0
    for r := 0; r < m; r ++ {
        for c := 0; c < n; c ++ {
            if grid[r][c] > 0 {
                fish := dfs(r, c)
                if fish > maxFish {
                    maxFish = fish
                }
            }
        }
    }
    return maxFish
}

File: from-2601-to-2700/2658-maximum-number-of-fish-in-a-grid/solution.py
class Solution:
    def findMaxFish(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        visited = set()

        def dfs(x, y):
            if x < 0 or x >= m or y < 0 or y >= n or grid[x][y] == 0 or (x, y) in visited:
                return 0

            visited.add((x, y))
            tot = grid[x][y]
            for dx, dy in dirs:
                tot += dfs(x + dx, y + dy)

            return tot

        result = 0
        for row in range(m):
            for col in range(n):
                if grid[row][col] > 0 and (row, col) not in visited:
                    result = max(result, dfs(row, col))

        return result
